### PR TITLE
10 video

### PR DESCRIPTION
**Pull request by Claudia108**
These are the latest changes:
- add videos table
- add video link embed method in video model and tests
- add functionality to add, play, update and delete youtube videos 
- add controller tests for sites and videos.

Request for input: (I added the error messages inline)
There are 3 failures in my videos controller test: create (does increment count), update (does not find id for show path), delete (does not decrement count).
Same errors are in the Site controller test for create and update.
The error has to be in my tests because the feature tests pass and the web app works fine.
Any input is highly appreciated.

### CODE REVIEW

File: test/controllers/admin/sites_controller_test.rb [hunk 1]
```diff
@@ -1,12 +1,40 @@
...
+                    name: "My Site",
+                    longitude: "-38.654",
+                    latitude: "45.345",
+                    site_type: "public_site"}
+    end
+    assert_equal 'no description available', assigns(:site).description
+    assert_equal 'no description available', assigns(:site).trinominal
+    assert_redirected_to site_path(assigns(:site))
+  end
+
```
Comment (Claudia108) [inline test/controllers/admin/sites_controller_test.rb:29]: Admin::SitesControllerTest#test_should_create_site [/Users/claudia/Turing/3module/projects/TexArch/test/controllers/admin/sites_controller_test.rb:14]:
"Site.count" didn't change by 1.
Expected: 4
  Actual: 3

File: test/controllers/admin/sites_controller_test.rb [hunk 2]
```diff
@@ -1,12 +1,40 @@
...
+                                name: "new name",
+                                longitude: "-56.789",
+                                latitude: "54.678",
+                                site_type: @site.site_type }
     assert_response :success
-    assert site.has_attribute?("name")
-    assert site.has_attribute?("longitude")
+    assert_redirected_to site_path(assigns(:site))
   end
 end
```
Comment (Claudia108) [inline test/controllers/admin/sites_controller_test.rb:43]: 2) Error:
Admin::SitesControllerTest#test_should_update_site:
ActionController::UrlGenerationError: No route matches {:action=>"show", :controller=>"sites", :id=>nil} missing required keys: [:id]
    test/controllers/admin/sites_controller_test.rb:38:in `block in class:SitesControllerTest'

File: test/controllers/admin/videos_controller_test.rb [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  test "should create video" do
+    assert_difference("Video.count") do
+      post :create, video: {
+                      title: @video.title,
+                      description: @video.description,
+                      link: @video.link}
+    end
+    assert_redirected_to video_path(assigns(:video))
+  end
+
```
Comment (Claudia108) [inline test/controllers/admin/videos_controller_test.rb:26]: 4) Failure:
Admin::VideosControllerTest#test_should_create_video [/Users/claudia/Turing/3module/projects/TexArch/test/controllers/admin/videos_controller_test.rb:14]:
"Video.count" didn't change by 1.
Expected: 3
  Actual: 2

File: test/controllers/admin/videos_controller_test.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+  test "should destroy video" do
+    @video
+    assert_difference('Video.count', -1) do
+      delete :destroy, id: @video.id
+    end
+    assert_redirected_to videos_path
+  end
+end
```
Comment (Claudia108) [inline test/controllers/admin/videos_controller_test.rb:47]: 3) Failure:
Admin::VideosControllerTest#test_should_destroy_video [/Users/claudia/Turing/3module/projects/TexArch/test/controllers/admin/videos_controller_test.rb:39]:
"Video.count" didn't change by -1.
Expected: 1
  Actual: 2

File: test/controllers/admin/videos_controller_test.rb [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+  test "should update video" do
+    patch :update, id: @video.id, video: {
+                                title: @video.title,
+                                description: @video.description,
+                                link: @video.link }
+    assert_response :success
+    assert_redirected_to video_path(assigns(:video))
+  end
+
```
Comment (Claudia108) [inline test/controllers/admin/videos_controller_test.rb:39]: 5) Error:
Admin::VideosControllerTest#test_should_update_video:
ActionController::UrlGenerationError: No route matches {:action=>"show", :controller=>"videos", :id=>nil} missing required keys: [:id]
    test/controllers/admin/videos_controller_test.rb:34:in `block in class:VideosControllerTest'

**Comment by jwashke** [comment]
Hey Claudia, I'm trying to get this running locally to take a look at it, but I'm missing some environment variables, and don't know which ones I need. Might be something to add to the readme.

**Comment by Claudia108** [comment]
Interesting. I have multiple keys for apis and paper clip in my
application.yml. Do you think this is what you are missing?
Thank you for taking a look!

**Comment by jwashke** [comment]
Yeah if you could post the name of the variables in your application.yml that would be cool. You don't need to post the keys, I have an AWS account and what not but I need to know what keys I need and what they should be called in your application.

**Comment by Claudia108** [comment]
Ok, get it. these are all the variables from my application.yml:

SECRET_KEY_BASE

// for google oauth:

GOOGLE_CLIENT_ID
GOOGLE_CLIENT_SECRET

// for googlemaps.api:

google_api_key

// for paperclip:

aws_bucket: "texarch"
aws_access_key_id
aws_secret_access_key

I hope this is helpful. Thank you so much for doing this!

**Comment by jwashke** [comment]
Got it up and running this morning. A couple notes: 
- Migrating the database had an error, it looks like you define a name column in sites twice in your migration files.
  In create_sites

``` ruby
class CreateSites < ActiveRecord::Migration
  def change
    create_table :sites do |t|
      t.string :longitude
      t.string :latitude
      t.string :name
      t.string :type, default: "public"

      t.timestamps null: false
    end
  end
end
```

And in add name to site

``` ruby
class AddNameToSite < ActiveRecord::Migration
  def change
     add_column :sites, :name, :string
  end
end
```

I had to comment out the second add_column line to get the migrations to run

**Comment by jwashke** [comment]
For the test errors themselves, I believe its because your not stubbing an admin login as all the errors are admin controller tests.

I commented out the line 
before_action :require_admin in the admin base controller and was left with these errors

```
  1) Failure:
Admin::VideosControllerTest#test_should_update_video [/Users/Josh/turing/4_module/TexArch/test/controllers/admin/videos_controller_test.rb:33]:
Expected response to be a <success>, but was <302>


  2) Failure:
Admin::SitesControllerTest#test_should_create_site [/Users/Josh/turing/4_module/TexArch/test/controllers/admin/sites_controller_test.rb:22]:
Expected: "no description available"
  Actual: "no trinominal available"


  3) Failure:
Admin::SitesControllerTest#test_should_update_site [/Users/Josh/turing/4_module/TexArch/test/controllers/admin/sites_controller_test.rb:37]:
Expected response to be a <success>, but was <302>
```

Error number 1 and 3 I believe should be 302, you have assert_response :success, but success is for the status code 200 ok. Since you are redirecting in these paths (and are asserting you are redirecting you probably want to assert_response 302 or assert_response :redirect

You can read more about that [here](http://apidock.com/rails/ActionController/Assertions/ResponseAssertions/assert_response)

The last error, number 2 is from these two lines

``` ruby
    assert_equal 'no description available', assigns(:site).description
    assert_equal 'no description available', assigns(:site).trinominal
```

I'm wondering if this is just a copy-pasta error and am assuming the second line should be

``` ruby
    assert_equal 'no trinominal available', assigns(:site).trinominal
```

Since in your migration file that is the default line for the trinominal.

When I commented out the require admin before action, changed the two responses to assert_response :redirect, and changed the no description to no trinominal, the tests all passed. So I think the main issue is needing to stub an admin login for the tests. 

If thats not right let me know and I'm happy to take another look. Or let me know if you have further questions or anything else you want me to take a look at.

**Comment by Claudia108** [comment]
Wow, I am impressed. Thank you so much for your time and figuring out these
details. I will try out your suggestions and let you know how it goes.

Claudia

**Comment by jwashke** [comment]
Happy to help, let me know how it all worked out.

**Comment by Claudia108** [comment]
Hi Josh, 
yes, you were right. I had to stub out the admin in the controller tests. Thank you for taking the time and for the suggestions. They all pass now. Just pushed up the changes. 
Do you know what how I could best fix the doubling migration? 
That must have happened in one of these very tired moments...

**Pull request merged**